### PR TITLE
TP-333: Allow fetching of "latest" versions of a relation in a single query (with TP-371)

### PR DESCRIPTION
**Pull request by NDevox**
The data model within TaMaTo is intended to be versioned. That is to say
that any row within a table derived from TrackedModel holds some identifying
columns which are non-unique. If multiple rows have the same data within these
identifying columns they are seen as multiple versions of the same object.

However foreign keys must be made on the primary key field of a table. With
this being the case it is possible to create a row on one table (e.g. Table1)
and link it to a row in another table (e.g. Table2). This row in Table2 can
then be "updated" - adding a new row to the table and a new version of the
object. In this case the foreign key on Table1 may be considered stale, but
only within the date range of the new row that has been added on Table2.

To solve this problem a mechanism has been added to allow fetching of the
"latest" row representing the latest version of an object for all
relations pertaining to derivatives of TrackedModels.

In python this allows us to define a model like so:

```python3
class ExampleModel(TrackedModel):
    # must be a TrackedModel
    other_model = models.ForeignKey(OtherModel, on_delete=models.PROTECT)
```

And then access the latest version of the relation by simply adding _current
to the attribute name like so:

```python3
example_model = ExampleModel.objects.first()
example_model.other_model_current  # Gets the most recent version
```

If done like in the above example, a query is generated specifically to fetch
this latest version of the row. However another mechanism has been added at
the queryset level which enable this to be done in a single query. This is
the `with_latest_links` QuerySet method.

This method utilises the `VersionGroup` instance the `TrackedModel` is attached
to and its `current_version` key back to `TrackedModel`. In this case there
is a direct linear path of Foreign Key relationships leading to the most
recent row for each relationship an object has.

An advantage of this over other proposed solutions is that it can be analysed
recursively (technically it can follow the relationships over the entire database
if need be) whilst still following direct FK relationships. All of which is done
within a simple Django `select_related` clause.

### CODE REVIEW

File: commodities/models.py [hunk 1]
```diff
@@ -153,6 +167,31 @@ def has_measure_in_tree(self):
...
+        ):
+            new_valid_between = DateTimeTZRange(
+                valid_between.lower, new_valid_between.upper
+            )
+        if not new_valid_between.upper or (
+            valid_between.upper and new_valid_between.upper > valid_between.upper
+        ):
+            new_valid_between = DateTimeTZRange(
+                new_valid_between.lower, valid_between.upper
+            )
```
Comment (stuaxo) [inline commodities/models.py:185]: ^^ hopefully we have tests for each of these `if`s .. :)   [I haven't scrolled down yet]

File: commodities/tests/test_importer.py [hunk 2]
```diff
@@ -198,21 +276,42 @@ def test_goods_nomenclature_indent_importer_create_with_parent_high_indent(valid
...
-    assert (
-        db_indent.indented_goods_nomenclature.sid
-        == indent.indented_goods_nomenclature.sid
+
+@pytest.mark.parametrize("update_type", [UpdateType.UPDATE, UpdateType.DELETE])
+def test_goods_nomenclature_indent_importer_update_with_parent_high_indent(
+    valid_user, update_type, date_ranges
+):
+    parent_indent = None
+    for idx in range(1, 6):
```
Comment (stuaxo) [inline commodities/tests/test_importer.py:287]: A quick comment about what the range here is (same for other hardcoded numbers around here).

File: commodities/tests/test_importer.py [hunk 3]
```diff
@@ -198,21 +276,42 @@ def test_goods_nomenclature_indent_importer_create_with_parent_high_indent(valid
...
+        valid_between=date_ranges.adjacent_no_end,
+        update_type=update_type,
     )
-    assert db_indent_node.get_parent() == parent_indent
-    assert db_indent.valid_between.lower == indent.valid_between.lower
 
+    db_indent = make_and_get_indent(updated_indent, valid_user, depth=4)
 
-def test_goods_nomenclature_indent_importer_multiple_parents(valid_user, date_ranges):
+    compare_indent(updated_indent, db_indent, 4, 1, 6)
```
Comment (stuaxo) [inline commodities/tests/test_importer.py:309]: Docstring for this test, and a comment about these numbers would be nice.

File: commodities/tests/test_importer.py (hunk 2 above)
Comment (NDevox) [reply commodities/tests/test_importer.py:287]: Done

File: commodities/tests/test_importer.py (hunk 3 above)
Comment (NDevox) [reply commodities/tests/test_importer.py:309]: Switched over to using keyword args. Added a docstring

**Pull request closed without merging**